x64: fix sysenter syscalls

canarysnort01 opened the pull request:
The MCS-related macros were causing the build to fail when configured to
use the sysenter instruction. Additionally, some message registers
weren't explicitly being passed to the assembly block for
seL4_NBSendRecv in this configuration.

canarysnort01 commented:
Testing this was a pain with the version of gcc I am using which really, really doesn't want sysenter to be used on x64. It was easier for me to assemble the instruction myself than figure out which combination of command line arguments tells gcc to stop stabbing me in the back.

```
sysenter:
.byte 0x0F,0x34

sysexitq:
.byte 0x48,0x0F,0x35
```

kent-mcleod commented:
> Testing this was a pain with the version of gcc I am using which really, really doesn't want sysenter to be used on x64. It was easier for me to assemble the instruction myself than figure out which combination of command line arguments tells gcc to stop stabbing me in the back.

Is this the error you were getting? 
```
/tmp/tmp.LOo5Rf8dhR/kernel/src/arch/x86/64/c_traps.c: Assembler messages:
/tmp/tmp.LOo5Rf8dhR/kernel/src/arch/x86/64/c_traps.c:227: Error: unsupported instruction `sysexit'
/tmp/tmp.LOo5Rf8dhR/kernel/include/arch/x86/arch/64/mode/fastpath/fastpath.h:190: Error: unsupported instruction `sysexit'
/tmp/tmp.LOo5Rf8dhR/kernel/include/arch/x86/arch/64/mode/fastpath/fastpath.h:190: Error: unsupported instruction `sysexit'
```

kent-mcleod commented:
Thanks for this, sorry that it was ignored for so long.  Still haven't looked into why newer gcc versions don't compile sysexitq but it's still possible to use clang to validate the fix. The gcc error is a different issue so IMO this can be merged provided the DCO is added.

kent-mcleod commented:
cf113c619b14d1c44b4350e0ba45ba30ac6497e0 introduced the sysexitq instruction instead of `rex.w sysexit`.  gcc doesn't seem to support sysexitq.

kent-mcleod commented:
https://github.com/seL4/seL4/pull/843 should address the compilation issues

canarysnort01 commented:
Thanks for investigating and fixing the gcc issue @kent-mcleod! I rebased and retested this PR with sel4test.

```bash
mkdir build
cd build
../init-build.sh -DPLATFORM=x86_64 -DRELEASE=false -DSIMULATION=true -DMCS=true -DKernelSyscall=sysenter
ninja
./simulate
```

One of the MCS scheduler tests failed. I don't know why but since syscalls are working I don't think they were caused by this PR.

```
...

Starting test 103: SCHED0011
Running test SCHED0011 (Test scheduler accuracy)
        Error: Check diff(102222740) <= period_ns + 2 * (1000UL * 1000UL)(102000000) failed. at line 1026 of file sel4test/projects/sel4test/apps/sel4test-tests/src/tests/scheduler.c
        Error: Check diff(103218170) <= period_ns + 2 * (1000UL * 1000UL)(102000000) failed. at line 1026 of file sel4test/projects/sel4test/apps/sel4test-tests/src/tests/scheduler.c
        Error: Check diff(105003870) <= period_ns + 2 * (1000UL * 1000UL)(102000000) failed. at line 1026 of file sel4test/projects/sel4test/apps/sel4test-tests/src/tests/scheduler.c
        Error: Check diff(105351130) <= period_ns + 2 * (1000UL * 1000UL)(102000000) failed. at line 1026 of file sel4test/projects/sel4test/apps/sel4test-tests/src/tests/scheduler.c
        Error: Check diff(106037470) <= period_ns + 2 * (1000UL * 1000UL)(102000000) failed. at line 1026 of file sel4test/projects/sel4test/apps/sel4test-tests/src/tests/scheduler.c
Test SCHED0011 failed
        Failure: result == SUCCESS at line 291 of file sel4test/projects/sel4test/apps/sel4test-driver/src/testtypes.c
        Error: result == SUCCESS at line 217 of file sel4test/projects/sel4test/apps/sel4test-driver/src/main.c

...

Starting test 167: Test all tests ran
Test suite failed. 166/167 tests passed.
*** FAILURES DETECTED ***
```

I've added the DCO.

kent-mcleod commented:
> One of the MCS scheduler tests failed. I don't know why but since syscalls are working I don't think they were caused by this PR.

Yea, this should be ok.  The same failures happen with `KernelSyscall=syscall` too. MCS scheduler tests need a more cycle accurate processor than what qemu typically emulates so these tests are very flaky on qemu.
